src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java

```diff
@@ -0,0 +1,138 @@
+        .build();
+  }
+
+  public static AssignmentOperationExpr multiplyAssignmentWithExprs(
+      VariableExpr variableExpr, Expr valueExpr) {
+    return builder()
+        .setVariableExpr(variableExpr)
+        .setValueExpr(valueExpr)
+        .setOperatorKind(OperatorKind.ASSIGNMENT_MULTIPLY)
+        .build();
+  }
+
+  private static Builder builder() {
+    return new AutoValue_AssignmentOperationExpr.Builder();
+  }
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    // Private setter.
+    abstract Builder setVariableExpr(VariableExpr variableExpr);
+
+    // Private setter.
+    abstract Builder setValueExpr(Expr valueExpr);
+
+    // Private setter.
+    abstract Builder setOperatorKind(OperatorKind operator);
+
+    abstract AssignmentOperationExpr autoBuild();
+
+    private AssignmentOperationExpr build() {
+      AssignmentOperationExpr assignmentOperationExpr = autoBuild();
+      TypeNode lhsType = assignmentOperationExpr.variableExpr().variable().type();
+      TypeNode rhsType = assignmentOperationExpr.valueExpr().type();
+      OperatorKind operator = assignmentOperationExpr.operatorKind();
+
+      // Check if the variable exprs have been declared, if yes, throw error.
+      Preconditions.checkState(
+          !assignmentOperationExpr.variableExpr().isDecl(),
+          String.format(
+              "Variable `%s` should not be declaration in the variable expression.",
+              assignmentOperationExpr.variableExpr().variable().name()));
+
+      // errorMsg is type checking error message for operators.
+      final String errorMsg =
+          String.format(
+              "Assignment operator %s can not be applied to %s, %s.",
+              operator, lhsType.toString(), rhsType.toString());
+
+      // Check type for multiply and assignment operator (*=).
+      if (operator.equals(OperatorKind.ASSIGNMENT_MULTIPLY)) {
+        Preconditions.checkState(isValidMultiplyAssignmentType(lhsType, rhsType), errorMsg);
+      }
+
+      // Check type for XOR and assignment operator (^=).
+      // TODO(summerji): Complete the type-checking for ^= and unit test.
+      if (operator.equals(OperatorKind.ASSIGNMENT_XOR)) {
+        Preconditions.checkState(isValidXORAssignmentType(lhsType, rhsType), errorMsg);
+      }
+      return assignmentOperationExpr;
+    }
+
+    // isValidMultiplyAssignmentType validates the types for LHS variable expr and RHS expr.
+    // *= can be only applied on Primitive numeric type.
+    private boolean isValidMultiplyAssignmentType(TypeNode variableType, TypeNode valueType) {
+      // LHS is numeric type, RHS should be any numeric type or any numeric boxed type.
+      if (TypeNode.isNumericType(variableType) && !TypeNode.isBoxedType(variableType)) {
+        return TypeNode.isNumericType(valueType);
+      }
+      // LHS is integer boxed type, RHS should be any numeric type except long, float, double.
+      if (variableType.equals(TypeNode.INT_OBJECT)) {
```

Comment:
Would prefer `INT` for consistency with other type checks, same for `LONG`, `FLOAT`, `DOUBLE` below.

Reply:
SG. Updated.